server/service/cron_schedues_test.go

```diff
@@ -0,0 +1,62 @@
+package service
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/viewer"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/fleetdm/fleet/v4/server/mock"
+	"github.com/fleetdm/fleet/v4/server/ptr"
+	"github.com/fleetdm/fleet/v4/server/service/schedule"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCronSchedulesService(t *testing.T) {
+	ds := new(mock.Store)
+	locker := schedule.SetupMockLocker("test_sched", "id", time.Now().Add(-1*time.Hour))
+	statsStore := schedule.SetUpMockStatsStore("test_sched", fleet.CronStats{
+		ID:        1,
+		StatsType: fleet.CronStatsTypeScheduled,
+		Name:      "test_sched",
+		Instance:  "id",
+		CreatedAt: time.Now().Add(-1 * time.Hour),
+		UpdatedAt: time.Now().Add(-1 * time.Hour),
+		Status:    fleet.CronStatsStatusCompleted,
+	})
+	jobsDone := uint32(0)
+
+	svc, ctx := newTestService(t, ds, nil, nil, &TestServerOpts{StartSchedules: []TestScheduleStarterFunc{
+		func(ctx context.Context, ds fleet.Datastore) fleet.CronScheduleStarterFunc {
+			return func() (fleet.CronSchedule, error) {
+				s := schedule.New(
+					ctx, "test_sched", "id", 1*time.Second, locker, statsStore,
+					schedule.WithJob("test_job", func(ctx context.Context) error {
+						time.Sleep(100 * time.Millisecond)
```

Comment:
An alternative way to do this synchronization is to use channels. You would create a chan above, and then here you would do:

```
<-waitCh
```

This will hang, so below line 50 will start the schedule. You can then immediately do line 53. Then after, you do

```
close(waitCh)
```

So the schedule finishes. If you're worried that it might be too fast to not wait on anything, then you can add a waitGroup, that way the test runs as fast as possible, without needing to wait. This test as it is has the potential to be flaky in a slow machine (e.g. a CI runner that is running many things at once).

Reply:
Sure thing. I found the `time.Sleep` made the tests easier to read and reason about since I felt it most closely tracked the types of scenarios I wanted to test without adding cognitive load for the reader to digest the ins-and-outs of other channels. But if you think the possible speedup and flake issues are worth the trade off, I can certainly do it the other way.

Reply:
One issue though is that the schedule package requires a minimum interval of 1 second (because of MySQL truncates timestamps at 1 second)

Reply:
In this test, we are expecting three jobs to be completed. The first is triggered right after the start, then a regularly scheduled run fires at t+1s and another regularly scheduled run at t+2s.

Reply:
The test confirms that the triggered run doesn't alter the expected behavior for regularly scheduled runs.